woocommerce/wc-admin
# REST API: Add /reports/variations

## Methodology

* Add `variation_id` column to the `wp_wc_order_product_lookup` table.
* Group results of querying this table by `variation_id`.
* Add attribute data as part of `extended_product_info = true` query arg.

A sample response:

```js
[
    {
        "product_id": 12,
        "variation_id": 28,
        "items_sold": 2,
        "gross_revenue": 84,
        "orders_count": 1,
        "extended_info": {
            "name": "Hoodie",
            "price": 42,
            "image": "<img width=\"324\" height=\"324\" src=\"http://tangaroa.test/wp-content/uploads/2018/11/hoodie-2-324x324.jpg\" class=\"attachment-woocommerce_thumbnail size-woocommerce_thumbnail\" alt=\"\" srcset=\"http://tangaroa.test/wp-content/uploads/2018/11/hoodie-2-324x324.jpg 324w, http://tangaroa.test/wp-content/uploads/2018/11/hoodie-2-150x150.jpg 150w, http://tangaroa.test/wp-content/uploads/2018/11/hoodie-2-300x300.jpg 300w, http://tangaroa.test/wp-content/uploads/2018/11/hoodie-2-768x768.jpg 768w, http://tangaroa.test/wp-content/uploads/2018/11/hoodie-2-416x416.jpg 416w, http://tangaroa.test/wp-content/uploads/2018/11/hoodie-2-100x100.jpg 100w, http://tangaroa.test/wp-content/uploads/2018/11/hoodie-2.jpg 801w\" sizes=\"(max-width: 324px) 100vw, 324px\" />",
            "permalink": "http://tangaroa.test/product/hoodie/",
            "attributes": [
                {
                    "id": 1,
                    "name": "color",
                    "option": "Red"
                },
                {
                    "id": 0,
                    "name": "logo",
                    "option": "No"
                }
            ]
        },
        "_links": {
            "product": [
                {
                    "href": "http://tangaroa.test/wp-json/wc/v3/products/12"
                }
            ],
            "variation": [
                {
                    "href": "http://tangaroa.test/wp-json/wc/v3/products/12/variation/28"
                }
            ]
        }
    },
]
```

## Test

Using [Postman](https://www.getpostman.com/) with an API key generated by a test site, check the endpoint for correct responses.

1. Create a product(s) with variations.
1. Create completed orders for those variations.
1. Or, if you have already created these, you may need to `Rebuild reports` (under WC > Status > Tools).
1.  All variations: `localhost/wp-json/wc/v3/reports/variations`.
2. Extended info: `localhost/wp-json/wc/v3/reports/variations?extended_variation_info=true`.
3. Limit to several variations: `localhost/wp-json/wc/v3/reports/variations?extended_variation_info=true&variations=357,359`.
3. Limit to several products: `localhost/wp-json/wc/v3/reports/variations?extended_variation_info=true&products=47,49`.

Fixes https://github.com/woocommerce/wc-admin/issues/906
